pkg/router/crl/crl.go

```diff
@@ -0,0 +1,238 @@
+// crls, and updates any crls that have expired.
+//
+// updateCRLFile will only read `caBundleFilename`, but will overwrite `crlFilename` if any changes are necessary.
+func updateCRLFile(caBundleFilename, crlFilename string) (time.Time, error) {
+	var nextCRLUpdate time.Time
+	clientCAData, err := os.ReadFile(caBundleFilename)
+	if err != nil {
+		return nextCRLUpdate, err
+	}
+
+	var crls map[string]*pkix.CertificateList
+	var crlFileExists = false
+	if _, err := os.Stat(crlFilename); err == nil {
+		crlFileExists = true
+		crlData, err := os.ReadFile(crlFilename)
+		if err != nil {
+			return nextCRLUpdate, err
+		}
+		crls, err = buildCRLMap(crlData)
+		if err != nil {
+			return nextCRLUpdate, err
+		}
+	}
+	if crls == nil {
+		crls = make(map[string]*pkix.CertificateList)
+	}
+
+	var subjectKeyIds []string
+	updated := false
+	now := time.Now()
+	for len(clientCAData) > 0 {
+		block, data := pem.Decode(clientCAData)
+		if block == nil {
+			break
+		}
+		clientCAData = data
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return nextCRLUpdate, fmt.Errorf("client CA configmap has an invalid certificate: %w", err)
+		}
+		subjectKeyId := hex.EncodeToString(cert.SubjectKeyId)
+		if len(cert.CRLDistributionPoints) == 0 {
+			continue
+		}
+		if crl, ok := crls[subjectKeyId]; ok {
+			if crl.HasExpired(now) {
+				log.Info("certificate revocation list has expired", "subject key identifier", subjectKeyId, "next update", crl.TBSCertList.NextUpdate.Format(time.RFC3339))
+			} else {
+				subjectKeyIds = append(subjectKeyIds, subjectKeyId)
+				if (nextCRLUpdate.IsZero() || crl.TBSCertList.NextUpdate.Before(nextCRLUpdate)) && crl.TBSCertList.NextUpdate.After(now) {
+					nextCRLUpdate = crl.TBSCertList.NextUpdate
+				}
+				continue
+			}
+		}
+		log.Info("retrieving certificate revocation list", "subject key identifier", subjectKeyId)
+		if crl, err := getCRL(cert.CRLDistributionPoints, now); err != nil {
+			// Creating or updating the configmap with incomplete data would compromise security by potentially
+			// permitting revoked certificates.
+			return nextCRLUpdate, fmt.Errorf("failed to get certificate revocation list for certificate key %s: %w", subjectKeyId, err)
+		} else if crl != nil {
+			crls[subjectKeyId] = crl
+			subjectKeyIds = append(subjectKeyIds, subjectKeyId)
+			log.Info("new certificate revocation list", "subject key identifier", subjectKeyId, "next update", crl.TBSCertList.NextUpdate.Format(time.RFC3339))
+			if (nextCRLUpdate.IsZero() || crl.TBSCertList.NextUpdate.Before(nextCRLUpdate)) && crl.TBSCertList.NextUpdate.After(now) {
+				nextCRLUpdate = crl.TBSCertList.NextUpdate
+			}
+			updated = true
+		}
+	}
+
+	if len(subjectKeyIds) == 0 {
+		if !crlFileExists {
+			// Make sure the crl file exists, even if it's empty
+			return nextCRLUpdate, os.WriteFile(crlFilename, []byte(""), crlFilePermissions)
+		} else {
+			return nextCRLUpdate, nil
+		}
```

Comment:
If we reach this point and `subjectKeyIds` is empty, that means that we parsed the CAs, and none of them had a CRL distribution point, right?  So why not simply truncate `crlFilename` when `subjectKeyIds` is empty?
```suggestion
		nextCRLUpdate = time.Time{}
		return nextCRLUpdate, os.Truncate(crlFilename, 0)
```
Then `ManageCRLs` should take a zero value to mean that no update is needed until the CA is updated.  Maybe it would be better to return a Boolean from `updateCRLFile` to indicate if no automatic update is required because there are no CRLs to update.

Reply:
Correct; if `subjectKeyIds` is empty, none of the certs in the CA bundle specify any CRL distribution points. I wrote it this way to avoid extra writes if the CRL file already exists, so that the fsnotify watch won't be triggered unnecessarily. Do you know if `os.Truncate` will always overwrite the specified file, or if it will only overwrite it if it's longer than the specified length?